src/providers/pkcs11/capability_discovery.rs

```diff
@@ -0,0 +1,123 @@
+#![allow(unused, trivial_numeric_casts)]
+use super::Provider;
+use crate::authenticators::ApplicationName;
+use crate::providers::pkcs11::to_response_status;
+use cryptoki::types::mechanism::Mechanism;
+use cryptoki::types::mechanism::MechanismInfo;
+use cryptoki::types::mechanism::MechanismType;
+use cryptoki::types::Ulong;
+use log::trace;
+use parsec_interface::operations::can_do_crypto;
+use parsec_interface::operations::can_do_crypto::CheckType;
+use parsec_interface::operations::psa_algorithm::Algorithm;
+use parsec_interface::operations::psa_key_attributes::Attributes;
+use parsec_interface::operations::psa_key_attributes::EccFamily;
+use parsec_interface::operations::psa_key_attributes::Type;
+use parsec_interface::requests::ResponseStatus::{InvalidEncoding, PsaErrorNotSupported};
+use parsec_interface::requests::Result;
+use std::convert::TryFrom;
+use std::ops::Deref;
+
+impl Provider {
+    pub(super) fn can_do_crypto_internal(
+        &self,
+        _app_name: ApplicationName,
+        op: can_do_crypto::Operation,
+    ) -> Result<can_do_crypto::Result> {
+        let attributes = op.attributes;
+        let check_type = op.check_type;
+        let supported_ecc_family_sizes = [192, 224, 256, 384, 512];
+        match attributes.key_type {
+            Type::EccKeyPair {
+                curve_family: EccFamily::SecpR1,
+            }
+            | Type::EccPublicKey {
+                curve_family: EccFamily::SecpR1,
+            } => {
+                if !(supported_ecc_family_sizes.contains(&attributes.bits)) {
+                    return Err(PsaErrorNotSupported);
+                }
+            }
+            Type::RsaKeyPair | Type::RsaPublicKey => (),
+            _ => return Err(PsaErrorNotSupported),
+        }
+        if attributes.policy.permitted_algorithms != Algorithm::None {
```

Comment:
In the case it isn't `None`, maybe it would be good to call: `attributes.compatible_with_alg(attributes.policy.permitted_algorithms)?;` to make sure the algorithm is compatible with the key type. That will return `PsaErrorNotPermitted` which can be added as a possible return value from this operation in the contract. [See here](https://docs.rs/psa-crypto/0.9.0/psa_crypto/types/key/struct.Attributes.html#method.compatible_with_alg)